File: core/src/main/java/com/gumse/gui/Basics/TextBox.cpp
#include "TextBox.h"
#include "Essentials/Unicode.h"
#include <System/Output.h>
#include <System/MemoryManagement.h>

TextBox::TextBox(Gum::Unicode str, Gum::GUI::Font *font, vec2 pos, vec2 size)
{
	this->sType = "TextBox";
	this->vPos = pos;
	this->vSize = size;
	this->sTitle = str;
	this->iAlignment = Alignment::CENTER;

	pBackgroundBox = new Box(vec2(0,0), ivec2(100,100));
    pBackgroundBox->setSizeInPercent(true, true);
	pBackgroundBox->setColor(vec4(0.3, 0.3, 0.3, 1));
	addElement(pBackgroundBox);

	pText = new Text(str, font, ivec2(50, 50), 0);
    pText->setPositionInPercent(true, true);
    pText->setOrigin(pText->getSize() / 2.0f);
	addElement(pText);

	resize();
	reposition();
}

TextBox::~TextBox() 
{
	Gum::_delete(pBackgroundBox);
	Gum::_delete(pText);
	Gum::_delete(value);
}

void TextBox::update()
{
	updatechildren();
}

void TextBox::updateText()
{
	pText->applyStringChanges();
    pText->setOrigin(pText->getSize() / 2.0f);
    pText->reposition();
}


void TextBox::updateOnPosChange()
{
	pText->setRenderBox(bbox2i(vActualPos, vActualSize));
}

void TextBox::updateOnSizeChange()
{
	switch(iAlignment)
	{
		case Alignment::LEFT:
    		pText->setOrigin(ivec2(0, pText->getSize().y / 2.0f) + v2TextOffset);
    		pText->setPosition(ivec2(0,50));
			break;

		case Alignment::CENTER:
    		pText->setOrigin(pText->getSize() / 2.0f + v2TextOffset);
    		pText->setPosition(ivec2(50,50));
			break;

		case Alignment::RIGHT:
    		pText->setOrigin(pText->getSize() * vec2(1.0f, 0.5f) + v2TextOffset);
    		pText->setPosition(ivec2(100,50));
			break;
	};
	pText->setRenderBox(bbox2i(vActualPos, vActualSize));
    pText->reposition();
}

void TextBox::updateOnColorChange()
{
	this->pBackgroundBox->setColor(v4Color);
}

void TextBox::setString(Gum::Unicode str)
{
	pText->setString(str);
	updateText();
}

void TextBox::setAlignment(Alignment alignment) { this->iAlignment = alignment; updateOnSizeChange(); }
void TextBox::setTextOffset(ivec2 offset)		{ this->v2TextOffset = offset; updateOnSizeChange(); }
void TextBox::setTexture(Texture *tex) 			{ this->pBackgroundBox->setTexture(tex); }
void TextBox::setTextColor(vec4 color)      	{ this->pText->setColor(color); }
void TextBox::setTextSize(int size)				{ this->pText->setCharacterHeight(size); }
void TextBox::setMaxTextlength(int length)		{ this->pText->setMaxLength(length); }

ivec2 TextBox::getTextSize()					{ return this->pText->getSize(); }
bool TextBox::isMouseInside() 					{ return this->pBackgroundBox->isMouseInside(); }
Texture* TextBox::getTexture() 					{ return this->pBackgroundBox->getTexture(); }
Text* TextBox::getText()						{ return this->pText; }
ivec2 TextBox::getTextOffset()					{ return this->v2TextOffset; }
Box* TextBox::getBox()							{ return this->pBackgroundBox; }
Gum::Unicode& TextBox::getString()				{ return this->pText->getString(); }

File: core/src/main/java/com/gumse/gui/Basics/TextBox.java
package com.gumse.gui.Basics;

import com.gumse.gui.Locale;
import com.gumse.gui.Font.Font;
import com.gumse.gui.Font.FontManager;
import com.gumse.gui.Primitives.Box;
import com.gumse.gui.Primitives.RenderGUI;
import com.gumse.gui.Primitives.Text;
import com.gumse.gui.XML.XMLGUI.XMLGUICreator;
import com.gumse.maths.*;
import com.gumse.system.filesystem.XML.XMLNode;
import com.gumse.textures.Texture;
import com.gumse.tools.Output;

public class TextBox extends RenderGUI
{
	public enum Alignment {
		LEFT,
		CENTER,
		RIGHT,
	};

	
    private Box	pBackgroundBox;
	private Text pText;
    private String sActualText, sFinalText;
	private Alignment iAlignment;
	private ivec2 v2TextOffset;
    private boolean bAutoInsertLinebreaks;
    private int iNewlineInsertOffset;


	public TextBox(String str, Font font, ivec2 pos, ivec2 size)
	{
		this.sType = "TextBox";
		this.vPos.set(pos);
		this.vSize.set(size);
		this.sTitle = str;
		this.iAlignment = Alignment.CENTER;
		this.v2TextOffset = new ivec2(0,0);
        this.bAutoInsertLinebreaks = false;
        this.iNewlineInsertOffset = 0;

		pBackgroundBox = new Box(new ivec2(0,0), new ivec2(100,100));
		pBackgroundBox.setSizeInPercent(true, true);
		addElement(pBackgroundBox);

		pText = new Text(str, font, new ivec2(50, 50), 0);
		pText.setPositionInPercent(true, true);
		pText.setOrigin(ivec2.div(pText.getSize(), 2.0f));
		pText.setCharacterHeight((int)(size.y * 0.9));
		addElement(pText);

        setString(str);

		resize();
		reposition();
	}

	public void cleanup() 
	{
		/*Gum::_delete(pBackgroundBox);
		Gum::_delete(pText);
		Gum::_delete(value);*/
	}

	public void updateText()
	{
		updateOnSizeChange();
	}


	protected void updateOnPosChange()
	{
		pText.setRenderBox(new bbox2i(vActualPos, vActualSize));
	}

	protected void updateOnSizeChange()
	{
        if(bAutoInsertLinebreaks)
        {
            insertLinebreaks();
            pText.setString(sFinalText);
        }

		ivec2 textSize = pText.getSize();
		switch(iAlignment)
		{
			case LEFT:
				pText.setOrigin(ivec2.add(new ivec2(0, (int)(textSize.y / 2.0f)), v2TextOffset));
				pText.setPosition(new ivec2(0,50));
				break;

			case CENTER:
				pText.setOrigin(ivec2.add(ivec2.div(textSize, 2.0f), v2TextOffset));
				pText.setPosition(new ivec2(50,50));
				break;

			case RIGHT:
				pText.setOrigin(ivec2.add(ivec2.mul(textSize, new vec2(1.0f, 0.5f)), v2TextOffset));
				pText.setPosition(new ivec2(100,50));
				break;
		};
		//pText.setOrigin(new ivec2(0, 0));
		//pText.getOrigin().print();
		pText.setRenderBox(new bbox2i(vActualPos, vActualSize));
		pText.reposition();
		//ivec2.sub(getPosition(), pText.getPosition()).print();
	}

	@Override
	protected void updateOnColorChange()
	{
		this.pBackgroundBox.setColor(v4Color);
	}

	@Override
	protected void updateOnThemeChange() {
        if(!sLocaleID.isEmpty())
			setString(Locale.getCurrentLocale().getString(sLocaleID));
	}

	public void setString(String str)
	{
        sActualText = str;
        if(bAutoInsertLinebreaks)
            insertLinebreaks();
        else
            sFinalText = sActualText;

		pText.setString(sFinalText);
		updateText();
	}

    private void insertLinebreaks()
    {
        int maxwidth = (int)(vActualSize.x * 0.85);
        int halfwidth = maxwidth / 2;
        int spacewidth = pText.getStringSize(" ").x * 3;
        String retString = "";
        this.iNewlineInsertOffset = 0;

        String[] words = sActualText.split(" ");
        for(int i = 0; i < words.length; i++)
        {
            String word = words[i];
            int wordwidth = pText.getStringSize(word).x + spacewidth;
            iNewlineInsertOffset += wordwidth;

            //Output.info(word + " " + wordwidth + " " + iNewlineInsertOffset + " " + maxwidth);
            if(wordwidth >= halfwidth)
            {
                retString += splitWord(word, maxwidth) + " ";
                continue;
            }
            else if(iNewlineInsertOffset >= maxwidth)
            {
                retString += "\n";
                iNewlineInsertOffset = 0;
            }

            retString += word + " ";
        }
        sFinalText = retString.toString();
    }

    private String splitWord(String word, int maxwidth)
    {
        StringBuilder retStr = new StringBuilder(word);
        //int offset = startoffset;
        for(int i = 0; i < word.length(); i++)
        {
            char currentChar = word.charAt(i);
            iNewlineInsertOffset += pText.getCharSize(currentChar).x;

            if(iNewlineInsertOffset >= maxwidth)
            {
                retStr.setCharAt(i, '\n');
                retStr.insert(i, currentChar);
                    
                iNewlineInsertOffset = 0;
            }
        }

        return retStr.toString();
    }

	public void setAlignment(Alignment alignment)    { this.iAlignment = alignment; updateOnSizeChange(); }
	public void setTextOffset(ivec2 offset)		     { this.v2TextOffset = offset; updateOnSizeChange(); }
	public void setTexture(Texture tex) 		     { this.pBackgroundBox.setTexture(tex); }
	public void setTextColor(vec4 color)      	     { this.pText.setColor(color); }
	public void setTextSize(int size)			     { this.pText.setCharacterHeight(size); }
	public void setMaxTextlength(int length)	     { this.pText.setMaxLength(length); }
	public void setCornerRadius(vec4 radius)	     { this.pBackgroundBox.setCornerRadius(radius); }
    public void setAutoInsertLinebreaks(boolean ins) { this.bAutoInsertLinebreaks = ins; pText.setFadeOverride(ins, ins); }

	public ivec2 getTextSize()					  { return this.pText.getSize(); }
	public Texture getTexture() 				  { return this.pBackgroundBox.getTexture(); }
	public Text getText()						  { return this.pText; }
	public ivec2 getTextOffset()				  { return this.v2TextOffset; }
	public Box getBox()							  { return this.pBackgroundBox; }
	public String getString()				      { return this.pText.getString(); }

    public static XMLGUICreator createFromXMLNode() 
    {
        return (XMLNode node) -> { 
			String fontName = node.getAttribute("font");
			Font font = (!fontName.equals("") ? FontManager.getInstance().getFont(fontName) : FontManager.getInstance().getDefaultFont());
	
			int fontsize  = node.getIntAttribute("fontsize", 0);
			int maxlength = node.getIntAttribute("maxlength", 0);
			
			TextBox textboxgui = new TextBox(node.content, font, new ivec2(0,0), new ivec2(0,0));
			if(fontsize > 0)
				textboxgui.setTextSize(fontsize);
			
			textboxgui.getText().setMaxLength(maxlength);
			return textboxgui;
        };
    };
};